File: architecture-skills/object-oriented-programming-paradigm/examples/classes-and-objects/classes-and-objects.py
"""Classes & objects in Python.

A class bundles state (owner, balance) with the behavior that guards
it; each instance gets independent state via __init__. Class attributes
are shared across instances (the account counter), and per-instance
state must be assigned in __init__ — a class-level mutable would be
shared by every account. transfer() stays a module-level free function
because it belongs to no single account.
"""

from __future__ import annotations


class BankAccount:
    """Blueprint: every account has an owner, a balance, and a number."""

    _accounts_opened = 0  # class attribute: shared by all instances

    def __init__(self, owner: str, opening_balance: float = 0.0) -> None:
        if opening_balance < 0:
            raise ValueError("opening balance cannot be negative")
        BankAccount._accounts_opened += 1
        self.number = f"ACC-{BankAccount._accounts_opened:04d}"
        self.owner = owner            # instance attributes: one copy
        self.balance = opening_balance  # per object, never shared

    # Instance methods: functions with an implicit reference (self).
    def deposit(self, amount: float) -> None:
        if amount <= 0:
            raise ValueError("deposit must be positive")
        self.balance += amount

    def withdraw(self, amount: float) -> None:
        if amount > self.balance:
            raise ValueError(f"{self.number}: insufficient funds")
        self.balance -= amount

    def __repr__(self) -> str:
        return f"{self.number} ({self.owner}): ${self.balance:.2f}"

    # Class method: bound to the class, callable with zero instances.
    @classmethod
    def total_opened(cls) -> int:
        return cls._accounts_opened


def transfer(source: BankAccount, target: BankAccount, amount: float) -> None:
    """Free function: coordinates two objects, owns neither."""
    source.withdraw(amount)
    target.deposit(amount)


if __name__ == "__main__":
    # Two objects from one blueprint, each with independent state.
    alice = BankAccount("Alice", opening_balance=100.0)
    bob = BankAccount("Bob")

    alice.deposit(50.0)
    bob.deposit(20.0)
    alice.withdraw(30.0)

    print(alice)  # Alice's mutations never touched Bob's balance
    print(bob)

    transfer(alice, bob, 40.0)
    print(f"after transfer -> {alice} | {bob}")

    print(f"accounts opened: {BankAccount.total_opened()}")

File: architecture-skills/object-oriented-programming-paradigm/examples/classes-and-objects/classes-and-objects.php
<?php
// Classes & objects in PHP.
//
// PHP 8 constructor property promotion declares and assigns instance
// fields directly in the constructor signature; readonly locks them
// after construction. `static` properties/methods are the class-level
// members, and transfer() stays a plain namespaced function because it
// belongs to no single account.

declare(strict_types=1);

final class BankAccount
{
    // Static property: one counter shared by all instances.
    private static int $accountsOpened = 0;

    public readonly string $number;
    private float $balance;

    // Constructor: promotion declares $owner as an instance field.
    public function __construct(
        public readonly string $owner,
        float $openingBalance = 0.0,
    ) {
        if ($openingBalance < 0) {
            throw new InvalidArgumentException('opening balance cannot be negative');
        }
        $this->number = sprintf('ACC-%04d', ++self::$accountsOpened);
        $this->balance = $openingBalance;
    }

    // Instance methods: $this binds them to one object.
    public function deposit(float $amount): void
    {
        if ($amount <= 0) {
            throw new InvalidArgumentException('deposit must be positive');
        }
        $this->balance += $amount;
    }

    public function withdraw(float $amount): void
    {
        if ($amount > $this->balance) {
            throw new RuntimeException("{$this->number}: insufficient funds");
        }
        $this->balance -= $amount;
    }

    public function __toString(): string
    {
        return sprintf('%s (%s): $%.2f', $this->number, $this->owner, $this->balance);
    }

    // Static method: bound to the class, callable with no instance.
    public static function totalOpened(): int
    {
        return self::$accountsOpened;
    }
}

// Free function: coordinates two objects, owns neither.
function transfer(BankAccount $from, BankAccount $to, float $amount): void
{
    $from->withdraw($amount);
    $to->deposit($amount);
}

// Two objects from one blueprint, each with independent state.
$alice = new BankAccount('Alice', 100.0);
$bob = new BankAccount('Bob');

$alice->deposit(50.0);
$bob->deposit(20.0);
$alice->withdraw(30.0);

echo $alice, PHP_EOL; // Bob's balance is untouched
echo $bob, PHP_EOL;

transfer($alice, $bob, 40.0);
echo 'after transfer:', PHP_EOL;
echo $alice, PHP_EOL;
echo $bob, PHP_EOL;

echo 'accounts opened: ', BankAccount::totalOpened(), PHP_EOL;
